### DISCUSSION TITLE
i.MX8MP OpenAMP Cortex-M7

### DISCUSSION THREAD

**Discussion by lbsobolewski** [zephyrproject-rtos/zephyr#80011] [Q&A]
Hi @alxlastur ,
Thanks a lot for your work on OpenAMP for the i.MX8MP platform in the PR https://github.com/zephyrproject-rtos/zephyr/pull/78585.
I'm encoutering some issues while trying to use it on another board powered by the exact same SoC.

Could you please explain your testing methods and environment?
I'm especially interested in which Linux kernel version you are using and which OpenAMP/RPmsg sample you ran for your tests?
Could you also provide some logs of these samples run (both Zephyr and Linux logs if possible)?

On my side, I'm currently stuck at `rproc_virtio_get_status` step (towards the end of `lib/open-amp/open-amp/lib/remoteproc/remoteproc_virtio.c`) where my Zephyr firmware is not able to read the `VIRTIO_CONFIG_STATUS_DRIVER_OK` flag; it keeps reading 0 so it never returns.
I have already verified that Linux was writing the relevant flag (it writes the value 7) but wasn't able to verify that both Linux and Zephyr are looking at it at the same register address.

Thank you for your help

**Comment by alxlastur** [comment]
Hi @lbsobolewski 

We are working on an Application Note, possibly a Patch that should solve this. 
I will post the link to the AN when is done.

**Reply by lbsobolewski** [reply]
Great news! Do you have an idea of when the note will be published?
And could you tell us which kernel versions it will support? We're using Linux 5.15 and we thought that an incompatibility between the drivers from this Linux release and the latest Zephyr version could occur.

**Comment by sebd71** [comment]
Hello @alxlastur ,
I'm working with @lbsobolewski and we're still struggling with this problem.
Does you Application Note available somewhere now ?
If not, is it possible to have a preview of that document and/or patches ?
Thanks in advance for your help.

**Comment by dbaluta** [comment]
@sebd71  @lbsobolewski 

We are using linux kernel 6.6 kernel (we can have a look here: https://github.com/nxp-imx/linux-imx), branch lf-6.6.y.
With this patch applied: https://lore.kernel.org/imx/CAEnQRZDoYvK-YXLjqbXsRAWDkHrWNOoR1OCCWxs+AfNUDuPB_w@mail.gmail.com/T/


I guess this would work on your kernel also, you can just port it. 

We also have this workaround applied:

```
--- a/arch/arm64/boot/dts/freescale/imx8mp-evk-rpmsg.dts
+++ b/arch/arm64/boot/dts/freescale/imx8mp-evk-rpmsg.dts
@@ -121,7 +121,7 @@ imx8mp-cm7 {
                mboxes = <&mu 0 1
                          &mu 1 1
                          &mu 3 1>;
-               memory-region = <&vdevbuffer>, <&vdev0vring0>, <&vdev0vring1>, <&rsc_table>;
+               memory-region = <&vdevbuffer>, <&vdev0vring0>, <&vdev0vring1>;//, <&rsc_table>;
                status = "okay";
                fsl,startup-delay-ms = <500>;
        };
```

Then we also have some changes in ATF. I will let @alxlastur  comment on that.  He is also working on creating an App Note but that will take a while until it gets all necessary approvals to be published.

**Reply by alxlastur** [reply]
Hi,

For atf, on imx8mplus platform iti is important to know that Linux is using to start the Cortex M7 with remoteproc, it will do call to ATF to start the Cortex M7 in the gpc_common.c https://github.com/nxp-imx/imx-atf/blob/616a4588f333522d50a55bedd2b9a90a51474a75/plat/imx/imx8m/gpc_common.c#L382

Here at line 382-383 the ATF will configure the RDC module to restrict the access of Cortex A to TCM memory of Cortex M and because of this Cortex A won't be able to access the resource table that resides in TCM. So the user should take care of regaining the access of Cortex A to TCM memory. One quick way is to comment these 2 lines, butIn a real application the user should give access only to that specific address range in which the resource table is located.


Please let us know how the tests are working on your side.


Best regards,
Alex

**Reply by lbsobolewski** [reply]
Hello,

I tested all modifications separately and simultaneously (applying the patch to my kernel, removing `rsc_table` from the list of reserved memory regions in the device tree, and commenting the two lines in ATF), nothing seems to change: FreeRTOS demos still work, Zephyr demos still don't work.

Also I made a mistake in my previous message, we use a kernel version 5.10, not 5.15 (but that doesn't prevent FreeRTOS from working).

Thank you for your help, I'll wait for the AN to be published for further testing.

**Reply by alxlastur** [reply]
Hi @lbsobolewski ,

After you made all the changes did you still get stuck in rproc_virtio_get_status()  or is it in another part of the code?


Alex

**Reply by lbsobolewski** [reply]
Hi @alxlastur , it's still stuck at the same place yes.

**Comment by lbsobolewski** [comment]
Hello @alxlastur , have you made significant progress on the application note? Maybe you have a more accurate idea about when you'll be able to publish it?
Thanks

**Comment by Theodore-Bardy** [comment]
Hello, I'm not sure to fully understand the issue you're encountering. But if the Zephyr set the resource table at the wrong place, this can lead to infinite loop or errors.
To offset the resource table at the address it is supposed to be you can use the Kconfig option `CONFIG_OPENAMP_COPY_RSC_TABLE=y` and specify in your device tree overlay where the resource table is expected by Linux using the Zephyr device tree node `zephyr,ipc_rsc_table`, such as:
```
+               zephyr,ipc_rsc_table = &rsc_table;
                zephyr,ipc_shm = &shram;
                zephyr,ipc = &mailbox0;
        };

+       rsc_table: memory@550ff000 {
+               compatible = "mmio-sram";
+               reg = <0x550ff000 0x1000>;
+       };
+
        shram: memory@55000000 {
                compatible = "mmio-sram";
                reg = <0x55000000 0x500000>;
```

**Comment by pefech** [comment]
Hello everyone,
was able to use OpenAMP with RPMsg on i.MX8MP before. I was using Kernel v6.6.23 with the patch showed on this page:

https://phytec.github.io/doc-bsp-yocto/coprocessor/coprocessor.html#openamp-using-resource-table

It worked with the Zephyr OpenAMP samples and with this custom implementation here: https://github.com/phytec/zephyr-software-defined-hardware

I hope this helps in some way.
